File: wordsearch.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.row = len(board)
        self.column = len(board[0])
        self.flag= False
        if self.row==1 and self.column == 1 and board[0][0]==word:
            return True

        for i in range(self.row):
            for j in range(self.column):
                if not self.flag:
                    self.dfs(board,i,j,word,0)


        return self.flag

    def dfs(self,board,row,col,word,idx):
        if idx==len(word):
            self.flag= True
            return

        if row<0 or col<0 or row>=self.row or col>=self.column or board[row][col]=="#":
            return

        if board[row][col]!=word[idx]:
            return

        temp = board[row][col]
        board[row][col] = "#"

        if row >= 1:
            self.dfs(board, row - 1, col, word, idx + 1)
        if row < self.row - 1:
            self.dfs(board, row + 1, col, word, idx + 1)
        if col >= 1:
            self.dfs(board, row, col - 1, word, idx + 1)
        if col < self.column - 1:
            self.dfs(board, row, col + 1, word, idx + 1)

        board[row][col] = temp

        

        

File: wordsearch.java
class Solution {
    int m,n;
    int[][] dirs; 
    public boolean exist(char[][] board,String word) {
        if(board == null||board.length==0){
            return false;        
        }
        m = board.length;
        n = board[0].length;
        dirs = new int[][] {{-1,0},{1,0},{0,-1},{0,1}};//UDLR
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(board[i][j]==word.charAt(0)){
                   if(backtrack(board,word,0,i,j)){
                    return true;
                   }
                }
            }
        }
        return false;
    }
    private boolean backtrack(char[][] board, String word, int index, int row, int col){
        //base
        if(index==word.length()){
           return true;
        }
        if(row < 0 || row == m || col < 0 || col == n || board[row][col] == '#'){
            return false;
        }
        //logic
        if(board[row][col] == word.charAt(index)){
           //action
            char temp=board[row][col];
            board[row][col]='#';
            for(int[] dir:dirs){
                int nr=row+dir[0];
                int nc=col+dir[1];
                //recurse
                if(backtrack(board,word,index+1,nr,nc)){
                   return true;
                }
            }
            //backward
            board[row][col]=temp;
        }
        return false;     
    }
}